### PR TITLE
[FEATURE] Enhance hook provider

### PR DESCRIPTION
**Pull request by mk-mxp**
Split #443 as requested by @oliverklee.

I chose a merged array to be able to return complex data sets, as it is the most flexible solution I could think of. All hooks registered get the same parameters, and the resulting arrays are merged and returned to the caller.

I will add the CHANGELOG line when it is ready to merge.

### CODE REVIEW

**Comment by mk-mxp** [comment]
Ready for review.

File: Classes/Hooks/HookProvider.php [hunk 1]
```diff
@@ -86,27 +93,36 @@ public function __construct(string $interfaceName, string $index = '')
...
         }
     }
 
+    /**
+     * Executes the hooked-in methods, that return result arrays.
+     *
+     * @param string $method the method to execute
+     * @param mixed $params parameters to $method()
+     *
+     * @return array the merged result arrays
```
[the remark this answers is not in the document]
Comment (mk-mxp) [reply Classes/Hooks/HookProvider.php]: This is impossible, as the possible keys depend on the hook implementation itself (specified in the interface, if possible). It is also not required to have keys.

File: Classes/Hooks/HookProvider.php [hunk 2]
```diff
@@ -117,9 +117,11 @@ public function executeHookReturningMergedArray(string $method, ...$params): arr
 
         $result = [];
         foreach ($this->getHooks() as $hook) {
-            $result = array_merge($result, $hook->$method(...$params));
+            $result[] = $hook->$method(...$params);
         }
 
+        \array_merge(...$result);
```
Comment (oliverklee) [inline Classes/Hooks/HookProvider.php]: Shouldn't this be `$result = \array_merge(...$result);` or `return \array_merge(...$result);` (as the result of `array_merge` is passed as the return value)?

File: Classes/Hooks/HookProvider.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (oliverklee) [reply Classes/Hooks/HookProvider.php]: Then I propose something along the lines of `with numeric or string keys, may contain duplicate values` to make this clear to the reader.

**Pull request merged**